usr/share/rear/lib/framework-functions.sh

```diff
@@ -7,7 +7,61 @@
 #
 # convert tabs into 4 spaces with: expand --tabs=4 file >new-file
 
-# source a file given in $1
+# source only a trustworthy file given in $1
+# A file is considered trustworthy to be sourced
+# when its file owner is one of the TRUSTED_FILE_OWNERS.
+# Because only the file owner can 'chmod' (cf. "man 2 chmod": caller's EUID must match owner)
+# we can sufficiently safely assume that a file which is onwed by one of the TRUSTED_FILE_OWNERS
+# is sufficiently trustworthy to be sourced without further additional checks
+# (e.g. if other users have permissions to modify the file or special ACLs).
+# Furthermore it should not be ReaR's task to prevent TRUSTED_FILE_OWNERS
+# from doing what they want (a.k.a. "final power to the user")
+# or simply put: TRUSTED_FILE_OWNERS means we do trust them.
+function SourceTrustworthy () {
+    local source_file="$1"
+    local source_return_code=0
+    local source_file_owner_name=""
+    local saved_bash_flags_and_options_commands=""
+    # Ensure source file is a regular file (or a link to a regular file).
+    # I.e. skip non-regular files like directories, device nodes, or file not found.
+    # It also returns here when no source file was specified (i.e. when $1 is empty).
+    if ! test -f "$source_file" ; then
+        # Show the source file name separated as '$source_file' to make it clear when it is empty:
+        Debug "Skipped 'SourceTrustworthy' because '$source_file' is not a regular file"
+        return 1
+    fi
+    # Ensure source file owner is one of the TRUSTED_FILE_OWNERS
+    # cf. the TRUSTED_FILE_OWNERS related code in build/default/990_verify_rootfs.sh
+    if test "$TRUSTED_FILE_OWNERS" ; then
+        source_file_owner_name="$( stat -c %U "$source_file" )"
+        if ! IsInArray "$source_file_owner_name" "${TRUSTED_FILE_OWNERS[@]}" ; then
+            # When the source file is not sourced because its owner is not one of the TRUSTED_FILE_OWNERS
+            # we ensure the user is notified because this reason is likely unexpected by the user:
+            LogPrintError "Refused 'SourceTrustworthy $source_file' because file owner '$source_file_owner_name' is not in TRUSTED_FILE_OWNERS"
+            return 1
+        fi
+    fi
+    # Save the bash flags and options settings so we can restore them after sourcing the source file:
+    { saved_bash_flags_and_options_commands="$( get_bash_flags_and_options_commands )" ; } 2>>/dev/$DISPENSABLE_OUTPUT_DEV
+    # The actual work (source the source file):
+    source "$source_file"
+    source_return_code=$?
+    test "0" -eq "$source_return_code" || Debug "SourceTrustworthy function: 'source $source_file' returns $source_return_code"
+    # Restore the bash flags and options settings to what they have been before sourcing the source file:
+    { apply_bash_flags_and_options_commands "$saved_bash_flags_and_options_commands" ; } 2>>/dev/$DISPENSABLE_OUTPUT_DEV
```

Comment:
This was previously only active during debugscript mode, right? Now it happens always?

Reply:
Yes,
because it was a bug that bash flags and options settings
were not always restored to what they had been before:

Two reasons:

1.)
This had made things behave different in debugscript mode
compared to non-debugscript mode.
Because in practice there was no issue with that
(at least no user reported one and I didn't detect one)
it sufficiently reliably indicates that it is OK
to always restore them to what they had been before.

2.)
In general after sourcing a file ReaR's standard
bash flags and options settings should be restored
so that all the code after sourcing a file behaves
normally and consistently, e.g. thing about when
a file is sourced only under certain conditions
and that file modifies bash flags and options settings
then all the code after sourcing that file is affected
by the condition why that file was sourced.
Or simply put:
Code that needs different bash flags and options settings
than what ReaR's standard is should set what is needed
only where it is actually needed (i.e. "locally")
but other code must not inherit it.

Where currently 'shopt' is used:
```
# find usr/sbin/rear usr/share/rear/ -type f \
 | xargs grep 'shopt' | grep -v '. *#'

usr/sbin/rear:
    shopt -p | tr '\n' ';'
shopt -s nullglob extglob

usr/share/rear/skel/default/etc/scripts/system-setup:
shopt -s nullglob

usr/share/rear/format/USB/default/200_check_usb_layout.sh:
    shopt -s nocasematch

usr/share/rear/lib/rear-shell.bashrc:
shopt -s nullglob extglob
```